DM-30162: Add link to processing data tutorials (#517)
lsst/pipe_tasks: doc/lsst.pipe.tasks/deblending-flags-overview.rst

Datasets           isPatchInner  isTractInner  isIsolated  fromBlend  isDeblendedSource  isDeblendedModelSource  isPrimary
---------------------------  ------------  ------------  ----------  ---------  -----------------  ----------------------  ---------
      isolated sources            x             x            x                          x                                      x
parents with multiple peaks       x             x
 models of isolated sources       x             x            x                                               x
models of deblended sources       x             x                        x              x                    x                 x
        sky objects               x             x            x           x              x                    x
   patch overlap regions                        x            x           x              x                    x
   tract overlap regions          x                          x           x              x                    x

> ---------------------------  ------------  ------------  ----------  ---------  -----------------  ----------------------  ---------

Author: That's fine. I've never seen this implementation of RST list tables, but it does seem easier and more maintainable. I was using the format described at https://docutils.sourceforge.io/docs/user/rst/quickref.html#tables .

Deblender input
===============

> The input `mergeDet` catalog to the deblender contains a list of parent sources, each row consisting of information about the parent, a footprint (a boolean mask of pixels in the input image that were detected as part of the parent blend), and a peak catalog (a list of peak locations and central flux values for all detected peaks in the parent blend). From here things change depending on which deblender is used, as single visit deblending uses the single-band :py:module:`lsst.meas.deblender` while multi-band deblending uses :py:module:`lsst.meas.extensions.scarlet` and generates a slightly different set of outputs.

Reviewer: ```suggestion
The input ``mergeDet`` catalog to the deblender contains a list of parent sources, each row consisting of information about the parent, a footprint (a boolean mask of pixels in the input image that were detected as part of the parent blend), and a peak catalog (a list of peak locations and central flux values for all detected peaks in the parent blend). From here things change depending on which deblender is used, as single visit deblending uses the single-band :doc:`/modules/lsst.meas.deblender` while multi-band deblending uses :doc:`lsst.meas.extensions.scarlet` and generates a slightly different set of outputs.
```

Also 

> The input mergeDet catalog to the deblender [..]

Is mergeDet the name of an argument? to a method/function? Perhaps rephrase for clarity.

Author: Neither, `mergeDet` is the name of an input catalog from the Butler.

Single-band deblender
=====================

> Prior to the adoption of scarlet as the default deblender for coadds, all deblending was done with :py:class:`lsst.meas.deblender.SourceDeblendTask`, which is still the single visit (single band) deblender. The basic idea of the algorithm is that most galaxies roughly exhibit 180 degree symmetry and (at shallow depths) are only blended with one other object in most cases. So a symmetric template (numerically equivalent to `x = np.min([x, x[::-1]])`) is made for all of the sources in a blend that do not fit the PSF, and the flux in the image is re-apportioned to each source based on the ratio of the templates for each pixel. This algorithm can fail for non-symmetric sources and for any blends where a source is blended with neighbors on both sides, causing inaccuracies in the measurement of all three sources. As the depth of the images increases blending becomes more severe and the instances where this algorithm fails increases, which is why a different deblender is used for co-added images in multiple bands.

Reviewer: ```suggestion
Prior to the adoption of scarlet as the default deblender for coadds, all deblending was done with `lsst.meas.deblender.SourceDeblendTask`, which is still the single visit (single band) deblender. The basic idea of the algorithm is that most galaxies roughly exhibit 180 degree symmetry and (at shallow depths) are only blended with one other object in most cases. So a symmetric template (numerically equivalent to ``x = np.min([x, x[::-1]])``) is made for all of the sources in a blend that do not fit the PSF, and the flux in the image is re-apportioned to each source based on the ratio of the templates for each pixel. This algorithm can fail for non-symmetric sources and for any blends where a source is blended with neighbors on both sides, causing inaccuracies in the measurement of all three sources. As the depth of the images increases blending becomes more severe and the instances where this algorithm fails increases, which is why a different deblender is used for co-added images in multiple bands.
```

FYI, code literals use double-back ticks, which links to Python APIs use single backticks.

Because the templates generated by the deblender are used to weight the flux from the image, the total flux in the footprints is conserved. This means that for isolated sources, the template that would be created by `SourceDeblendTask` is irrelevant, as all of the flux in the footprint would be returned. The result is an output catalog (in each band) with all of the parent blends at the top followed by all of the children deblended from one of the parents. So when this was the only deblender, selecting a set of unique sources was easy, you just cut on `deblend_nChild == 0`. This selected all of the isolated sources (from the parent section) and all of the deblended child sources.

Multi-band deblender (`scarlet`)
================================

> :py:class:`lsst.meas.extensions.scarlet.ScarletDeblendTask` is different, as it uses `scarlet` to create a *model* for each source in a blend. This is a philosophically different object, as there is no longer an assumption that all of the flux in the input image will be modeled by one of the children in the blend (this may change in the near future, but this is the current implementation). The results of the scarlet deblender will be biased by the assumptions that went into making the models, so it was decided that it would be a good idea to (by default) also model all of the isolated sources. This will allow comparisons of scarlet models of isolated sources to the un-modeled isolated source measurements to investigate the biases that scarlet is introducing and also gives users the option to choose between the un-modeled (parent) isolated source records and the scarlet model version of each isolated source. However this flexibility forced a change in the way that we select unique objects in a source catalog.

Reviewer: FYI, ScarletDeblendTasks, and probably other tasks related to deblending, need documentation pages. See https://developer.lsst.io/stack/task-topic-type.html No need to do that in this ticket, but keep that in mind for future documentation work.

Author: Created DM-30535.

> :py:class:`lsst.meas.extensions.scarlet.ScarletDeblendTask` is different, as it uses `scarlet` to create a *model* for each source in a blend. This is a philosophically different object, as there is no longer an assumption that all of the flux in the input image will be modeled by one of the children in the blend (this may change in the near future, but this is the current implementation). The results of the scarlet deblender will be biased by the assumptions that went into making the models, so it was decided that it would be a good idea to (by default) also model all of the isolated sources. This will allow comparisons of scarlet models of isolated sources to the un-modeled isolated source measurements to investigate the biases that scarlet is introducing and also gives users the option to choose between the un-modeled (parent) isolated source records and the scarlet model version of each isolated source. However this flexibility forced a change in the way that we select unique objects in a source catalog.

Reviewer: ```suggestion
`lsst.meas.extensions.scarlet.ScarletDeblendTask` is different, as it uses `scarlet <https://pmelchior.github.io/scarlet/>`__ to create a *model* for each source in a blend. This is a philosophically different object, as there is no longer an assumption that all of the flux in the input image will be modeled by one of the children in the blend (this may change in the near future, but this is the current implementation). The results of the scarlet deblender will be biased by the assumptions that went into making the models, so it was decided that it would be a good idea to (by default) also model all of the isolated sources. This will allow comparisons of scarlet models of isolated sources to the un-modeled isolated source measurements to investigate the biases that scarlet is introducing and also gives users the option to choose between the un-modeled (parent) isolated source records and the scarlet model version of each isolated source. However this flexibility forced a change in the way that we select unique objects in a source catalog.
```

("scarlet" is external documentation, so we need to link to it)

Flags set by the deblender
==========================

> Before we get into the flags set by pipe_tasks it is useful to understand the flags that are set in `SourceDeblendTask` and `ScarletDeblendTask` that relate to source selection.

Reviewer: > Before we get into the flags set by `pipe_tasks`

A bunch of things:

`pipe_tasks` is an EUPS package, and that concept isn't relevant here. Usually, we replace `pipe_tasks` with `lsst.pipe.tasks` in end-user documentation because that's what users actually care about. But I'm also a big confused because this is a super-generic comment. Perhaps re-write this to suggest what category of tasks you're referring to?

Reviewer: Also, a general comment that you should [**never** use "we" in end-user documentation](https://developers.google.com/style/person). Always write in the second person ("you..."). I'd suggest doing a search for "we" and re-writing those sentences.

- `parent`: the id in the catalog for the parent of this source record. This is actually set pre-deblender, where all top level records have `parent=0`.
- `deblend_nPeaks`: the number of peaks contained in the sources footprint.
- `deblend_nChild`: the number of peaks deblended by the deblender from this source and created as new source records in the catalog. This is different from `deblend_nPeaks` in that isolated sources that are not deblended by `SourceDeblendTask` and child peaks that were culled during deblending are not included in this count.
- `deblend_parentNPeaks`: The number of peaks contained in the parent of this source record.

> - `deblend_parentNChild`: the number of children deblended from the parent of this source record.

Reviewer: Again, everywhere you're using single backticks here, reStructuredText needs double backticks.

`isPrimary` and other flags added in `pipe_tasks`

> =================================================

Reviewer: Again, `pipe_tasks` is an EUPS package. Please rewrite this heading to clarify what tasks you're talking about.

Reviewer: Also, don't use code literals in headings (so drop the backticks; 
Also, we don't format headings (they should be plain text). Drop the backticks. If you have an underscore, you can escape it as `\_`.

> In addition to source records for deblended parents and multiple entries for isolated sources, output catalogs are also not unique because they may contain "pseduo" sources (eg. sky objects that have been added to assist with calibration but are not output sources) and, if the analysis is done over multiple patches and/or tracts, sources in the overlap region can exist in multiple overlapping patches (but always on the interior of only one). For this reason the :py:class:`lsst.pipe.tasks.SetPrimaryFlagsTask` sets a number of useful flags to assist users in determining a unique output catalog for their analysis.

Reviewer: ```suggestion
In addition to source records for deblended parents and multiple entries for isolated sources, output catalogs are also not unique because they may contain "pseduo" sources (eg. sky objects that have been added to assist with calibration but are not output sources) and, if the analysis is done over multiple patches and/or tracts, sources in the overlap region can exist in multiple overlapping patches (but always on the interior of only one). For this reason the :lsst-task:`~lsst.pipe.tasks.SetPrimaryFlagsTask` task sets a number of useful flags to help you determine a unique output catalog for your analysis.
```

- Using the `lsst-task` role to link to the task reference page
- Fixing the third-person tense.

> `detect_isPatchInner` and `detect_isTractInner`
> -----------------------------------------------

Reviewer: ```suggestion
detect\_isPatchInner and detect\_isTractInner
---------------------------------------------
```

`True` when:

- A source is in the inner region of a patch
- A source is in the inner region of a tract

Details

> ^^^^^^^

Reviewer: I recommend dropping all "Details" subheaders here because it seems to add noise and doesn't help with scanning down the page. (If you had more than a Details subheader for each flag it would be a different story)

or

- A source is flagged as `merge_peak_sky` in a `mergeDet` coadd catalog.

Details
^^^^^^^

`sky_source` is a flag in a single visit catalog to mark sky objects while `merge_peak_sky` is the coadd version (which states that a source was a sky object in at least one band). Any sources with either of these flags set should be ignored in a final source catalog as they are not astrophysical objects.

`detect_isIsolated`
-------------------

`True` when:

- A source only has a single peak (`deblend_nPeaks == 1`)
- A source is a top level parent (`parent == 0`) or its parent only had a single peak (`deblend_parentNPeaks == 1`)

Details
^^^^^^^

The `detect_isIsolated` flag marks sources that are not contained in a blend. This covers both isolated sources that are not modeled by the deblender (parents) and (in cases where the multi-band deblender is used) scarlet models of the isolated sources. Note that cutting on this flag will *not* give a unique set of sources, but can be useful for selecting all of the isolated sources to analyze the differences between measurements made on scarlet models and measurements made on the same isolated sources.

`detect_fromBlend`

> ^^^^^^^^^^^^^^^^^^

Reviewer: Change this heading level to `-`
